TelemetryDesignDoc.md

```diff
@@ -0,0 +1,327 @@
+
+The unified replacement API will look something like:
+
+```java
+// Simple imperative logging directly delegates to NT4
+Telemetry.publishDouble("key", value);
+// Can support dimensions through the API by delegating to metadata (C++ could use templates and be smarter)
+Telemetry.publishQuantity("key", value, "meters");
+
+// Declarative binding with suppliers/consumers
+Telemetry.publishDouble("key", () -> value);
+Telemetry.subscribeDouble("key", (double value) -> doThing(value));
+
+// Declarative binding of mutable classes similar to `putData(Sendable)` in existing impl
+// Optional final metadata string for aggregation
+var telemetryBinding = Telemetry.bind("key", telemetryNode);
+// Unbind later if needed (for cleanup etc)
+telemetryBinding.unbind();
+```
+
+Metadata for a given topic will be deep-merged into the existing metadata as they are introduced.  There does not seem to be any need for a system to remove metadata - such an interaction would be a code smell.
+
+### Replace Sendable with new, cleaner API
+
+`Sendable` should be replaced by a new interface, `TelemetryNode`.  The API shape for this will be fundamentally similar to `Sendable`:
+
+```java
+public interface TelemetryNode {
+  void bind(TelemetryBuilder builder);
+}
+```
+
+The "bind" method fulfills the role of `initSendable`, with a few key differences:
+
+Unlike `Sendable`, both `TelemetryNode` will support hierarchical nesting of implementing objects - that is, in addition to exposing binding handles for fields of primitive data type, it will allow *entire* `TelemetryNode`s to be added as children:
+
+```java
+@Override
+public void bind(TelemetryBuilder builder) {
+  // Add a `TelemetryNode` as a child; similarly-shaped API to ordinary binding
+  builder.bind("child", childNode);
+}
+```
+
+No auto-registration will be supported.  `TelemetryNode`s will only have their listeners bound if explicitly bound, either as a root node or as a child.
+
+Publishing and subscribing will no longer be overloaded onto a single `addProperty` method of the builder, since they only rarely overlap in practice:
+
+```java
+@Override
+public void bind(TelemetryBuilder builder) {
+  // API here matches the ordinary Telemetry API
+  builder
+    .publishDouble("publishedValue", () -> value)
+    .subscribeDouble("subscribedValue", (double value) -> { doThing(double); });
+}
+```
+
+#### Breaking out of the object tree
+
+It's sometimes useful to bind fields or children to paths that do not necessarily follow the in-code object tree - for example, to put certain data under a shared "diagnostics" aggregation even when it is held across different classes in code.
+
+To do this, users can call `bind`, `publish`, or `subscribe` directly on `Telemetry` instead of on the `TelemetryBuilder`.  The APIs will be identically-shaped, and will differ only in the property path locations.
```

Comment:
My team hasn't used Command-Based but will next season so I don't know the "right" way to do things. But a shared sensor might be GYRO that is used in autonomous to drive to a heading and in TELEOP is used to flip the joystick for field-orientation. Maybe these are both the drive train class but maybe not. Structure of code isn't a strong point of the students but it seems important for Commands and for this discussion of telemetry. This has to be easy to use by  mentors who are barely surviving herding a dozen students who don't know much and as we know don't read much documentation.